File: src/2024/day04/2/main.go
package main

import (
	_ "embed"
	"fmt"
	"strings"
)

//go:embed data.txt
var fileContents string

func readMatrix() [][]rune {
	lines := strings.Split(strings.TrimSpace(fileContents), "\n")
	matrix := make([][]rune, len(lines))
	for i, line := range lines {
		matrix[i] = []rune(line)
	}
	return matrix
}

func find3x3Patterns(matrix [][]rune) int {
	validPatterns := [][]rune{
		[]rune("MSMS"),
		[]rune("SMSM"),
		[]rune("SSMM"),
		[]rune("MMSS"),
	}

	count := 0
	for i := 0; i <= len(matrix)-3; i++ {
		for j := 0; j <= len(matrix[0])-3; j++ {
			corners := []rune{
				matrix[i][j],
				matrix[i][j+2],
				matrix[i+2][j],
				matrix[i+2][j+2],
			}
			center := matrix[i+1][j+1]

			if center == 'A' {
				for _, pattern := range validPatterns {
					if cornersMatch(corners, pattern) {
						count++
						break
					}
				}
			}
		}
	}
	return count
}

func cornersMatch(corners, pattern []rune) bool {
	cornersStr := string(corners)
	patternStr := string(pattern)
	return cornersStr == patternStr
}

func main() {
	matrix := readMatrix()
	result := find3x3Patterns(matrix)
	fmt.Println(result)
}

File: src/2024/day04/2/main.rs
use regex::Regex;
use std::fs;
use std::path::Path;

fn read_matrix() -> Vec<Vec<char>> {
    let current_file = Path::new(file!());
    let current_dir = current_file
        .parent()
        .expect("Failed to get parent directory");
    let file_path = current_dir.join("data.txt");

    fs::read_to_string(file_path)
        .expect("Failed to read file")
        .lines()
        .map(|line| line.chars().collect())
        .collect()
}

fn find_xs_count(line: &str) -> usize {
    line.matches("XMAS").count() + line.matches("SAMX").count()
}

fn find_3x3_patterns(matrix: &Vec<Vec<char>>) -> usize {
    let valid_patterns = vec![
        vec!['M', 'S', 'M', 'S'],
        vec!['S', 'M', 'S', 'M'],
        vec!['S', 'S', 'M', 'M'],
        vec!['M', 'M', 'S', 'S'],
    ];

    let mut count = 0;
    for i in 0..=matrix.len() - 3 {
        for j in 0..=matrix[0].len() - 3 {
            let corners = vec![
                matrix[i][j],
                matrix[i][j + 2],
                matrix[i + 2][j],
                matrix[i + 2][j + 2],
            ];
            let center = matrix[i + 1][j + 1];

            if center == 'A' {
                for pattern in &valid_patterns {
                    if corners == *pattern {
                        count += 1;
                        break;
                    }
                }
            }
        }
    }
    count
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matrix = read_matrix();
    let result = find_3x3_patterns(&matrix);
    println!("{}", result);
    Ok(())
}
